heron/statefulstorages/src/java/org/apache/heron/statefulstorage/localfs/LocalFileSystemStorage.java

```diff
@@ -134,6 +146,23 @@ public void dispose(String topologyName, String oldestCheckpointPreserved,
     }
   }
 
+  private void cleanCheckpoints(String path, int remaining) throws StatefulStorageException {
+    if (FileUtils.isDirectoryExists(path) && FileUtils.hasChildren(path)) {
+      File[] children = new File(path).listFiles();
+      Arrays.sort(children, Comparator.reverseOrder());
```

Comment:
I would suggest to break the logic into two steps:
1. generate the toDelete list.
2. for each one in the list, deleteDir().

The benefit is that you can write a unit test for step 1 easily to make sure the order is correct.